### ISSUE TITLE
Receiving data over Unix Datagram socket fails in prim_inet

### ISSUE THREAD

**Issue by hauleth** [erlang/otp#7605]
**Describe the bug**

When I try to receive message using `gen_udp:recv/2` on Unix Datagram socket it fails with error:

```
** exception error: no function clause matching prim_inet:split(0,<<"foo">>,"s") 
     in function  prim_inet:get_addr/2 
     in call from prim_inet:recvfrom0/3 
```

**To Reproduce**

1. Open receiving Unix Datagram socket using `gen_udp` with `{active, false}` mode.
2. Open sending Unix Datgagram socket using `gen_udp`.
3. Send message from socket in step 2. to socket in step 1.
4. Try to `gen_tcp:recv(S, 0)`

```
Erlang/OTP 25 [erts-13.2.2.1] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1]

Eshell V13.2.2.1  (abort with ^G)
1> {ok,S} = gen_udp:open(0, [{ip, {local, "socket"}}, binary, {active, false}]).                        
{ok,#Port<0.10>}
2> {ok,A} = gen_udp:open(0, [{ip, {local, "s"}}, binary, local, {active, false}]).                      
{ok,#Port<0.11>}
3> gen_udp:send(A, {local, "socket"}, "foo").                                                           
ok
4> gen_udp:recv(S, 0).
** exception error: no function clause matching prim_inet:split(0,<<"foo">>,"s") 
     in function  prim_inet:get_addr/2 
     in call from prim_inet:recvfrom0/3 
```

**Expected behavior**

Erlang/OTP 25 [erts-13.2.2.1] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1]

```
Eshell V13.2.2.1  (abort with ^G)
1> {ok,S} = gen_udp:open(0, [{ip, {local, "socket"}}, binary, {active, false}]).                        
{ok,#Port<0.10>}
2> {ok,A} = gen_udp:open(0, [{ip, {local, "s"}}, binary, local, {active, false}]).                      
{ok,#Port<0.11>}
3> gen_udp:send(A, {local, "socket"}, "foo").                                                           
ok
4> gen_udp:recv(S, 0).
{ok, <<"foo">>}
```

**Affected versions**

Tested only on OTP 25 (ERTS 13.2.2.1)

**Additional context**

Tested the same thing using `socket` as a receiving side by doing

```erl
{ok, S} = socket:open(local, dgram),
socket:bind(S, #{family => local, path => "socket"}),
% Send message from other socket
socket:recv(S).
```

And it works as expected.

**Comment by hauleth** [comment]
Ok, I have narrowed it down, and it happens when the socket is set to `binary`, if it is in default (`list`) mode then everything works.

**Comment by hauleth** [comment]
It seems that `split/2` in `prim_inet` accepts only lists as input, while if there is `{mode, binary}` then it will be feed with binary. I haven't tested it yet, but:

```diff
diff --git i/erts/preloaded/src/prim_inet.erl w/erts/preloaded/src/prim_inet.erl
index 89ee3f150a..3298f22be8 100644
--- i/erts/preloaded/src/prim_inet.erl
+++ w/erts/preloaded/src/prim_inet.erl
@@ -2771,10 +2771,14 @@ rev(L) -> rev(L,[]).
 rev([C|L],Acc) -> rev(L,[C|Acc]);
 rev([],Acc) -> Acc.
 
+split(N, B) when is_binary(B) -> bin_split(N, B, byte_size(B));
 split(N, L) -> split(N, L, []).
 split(0, L, R) when is_list(L) -> {rev(R),L};
 split(N, [H|T], R) when is_integer(N), N > 0 -> split(N-1, T, [H|R]).
 
+bin_split(N, B, Size) when N >= Size -> {B, <<>>};
+bin_split(N, B, Size) -> {binary_part(B, 0, N), binary_part(B, N + 1, Size - N)}.
+
 len(L, N) -> len(L, N, 0).
 len([], N, C) when is_integer(N), N >= 0 -> C;
 len(L, 0, _) when is_list(L) -> undefined;
```

Should do the job.

**Comment by RaimoNiskanen** [comment]
The fix should be as simple as this:
``` erlang
---------------------- erts/preloaded/src/prim_inet.erl -----------------------
index 89ee3f150a..a9264d551a 100644
@@ -2772,7 +2772,7 @@ rev([C|L],Acc) -> rev(L,[C|Acc]);
 rev([],Acc) -> Acc.
 
 split(N, L) -> split(N, L, []).
-split(0, L, R) when is_list(L) -> {rev(R),L};
+split(0, L, R) when is_list(L); is_binary(L) -> {rev(R),L};
 split(N, [H|T], R) when is_integer(N), N > 0 -> split(N-1, T, [H|R]).
 
 len(L, N) -> len(L, N, 0).

```
Data is sent from inet_drv.c as `{inet_async, S, Ref, {ok,[H1,...Hsz | Data]}}` where `Data`, when the socket is in binary mode, is a binary (in the tail of the list).

See PR #7623.

**Comment by RaimoNiskanen** [comment]
Fix merged to 'maint' to be released in OTP-26.0.
Thank you for reporting this issue!
